modules/nextflow/src/main/groovy/nextflow/cache/PathCacheStore.groovy

```diff
@@ -0,0 +1,170 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package nextflow.cache
+
+import java.nio.file.NoSuchFileException
+import java.nio.file.Path
+
+import com.google.common.hash.HashCode
+import groovy.transform.CompileStatic
+import nextflow.cache.CacheStore
+import nextflow.exception.AbortOperationException
+import nextflow.util.CacheHelper
+/**
+ * Implements the path-based cache store
+ *
+ * @author Ben Sherman <bentshermann@gmail.com>
+ */
+@CompileStatic
+class PathCacheStore implements CacheStore {
+
+    private final String LOCK_NAME = 'LOCK'
+
+    private final int KEY_SIZE
+
+    /** The session UUID */
+    private UUID uniqueId
+
+    /** The unique run name associated with this cache instance */
+    private String runName
+
+    /** The base path for the entire cache */
+    private Path basePath
+
+    /** The base path for this cache instance */
+    private Path dataPath
+
+    /** The lock file for this cache instance */
+    private Path lock
+
+    PathCacheStore(UUID uniqueId, String runName, Path basePath=null) {
+        this.KEY_SIZE = CacheHelper.hasher('x').hash().asBytes().size()
+        this.uniqueId = uniqueId
+        this.runName = runName
+        this.basePath = basePath ?: defaultBasePath()
+        this.dataPath = this.basePath.resolve("$uniqueId")
+        this.lock = dataPath.resolve(LOCK_NAME)
+    }
+
+    private Path defaultBasePath() {
+        final basePath = System.getenv('NXF_CACHE_PATH')
+        if( !basePath )
+            throw new IllegalArgumentException("NXF_CACHE_PATH must be defined when using the path-based cache store")
+
+        return basePath as Path
+    }
+
+    @Override
+    PathCacheStore open() {
+        acquireLock()
+        return this
+    }
+
+    @Override
+    PathCacheStore openForRead() {
+        if( !dataPath.exists() )
+            throw new AbortOperationException("Missing cache directory: $dataPath")
+        acquireLock()
+        return this
+    }
+
+    private void acquireLock() {
+        if( lock.exists() ) {
+            final msg = """
+                Unable to acquire lock for session with ID ${uniqueId}
+
+                Common reasons for this error are:
+                - You are trying to resume the execution of an already running pipeline
+                - A previous execution was abruptly interrupted, leaving the session open
+
+                You can see the name of the conflicting run by inspecting the contents of the following path: ${lock}
+                """
+            throw new IOException(msg)
+        }
+
+        lock.text = runName
+    }
+
+    @Override
+    void drop() {
+        dataPath.deleteDir()
+    }
+
+    @Override
+    void close() {
+        lock.delete()
+    }
+
+    @Override
+    void writeIndex(HashCode key, boolean cached) {
+        getCachePath(key).bytes = [ cached as byte ] as byte[]
+    }
```

Comment:
Not understanding this logic, the index file is essentially an append log, that keep the order of the execution of the tasks 

https://github.com/nextflow-io/nextflow/blob/s3-cache-store/modules/nextflow/src/main/groovy/nextflow/cache/DefaultCacheStore.groovy#L134-L137

Reply:
I think I did this because I wasn't sure if the index file and leveldb were independent, i.e. writing a cache entry to leveldb doesn't affect the index file. If they are independent, I will just save the index file as an S3 object and use it exactly as the default store.

Reply:
However, since I can't append to an S3 object, I will have to re-upload the entire index file every time a task is written. That should be okay for now, but ideally we would want to avoid that.

The index file iteration is used by the `clean` and `log` commands, but they don't need the order of execution. This is why I think we could discard the index file and instead simply list the directory of cache entries when we need to iterate.

Reply:
Yes, you can create an output stream with `Files.newOutputStream(Path)`

Reply:
So that does some kind of streaming upload that continues for the entire pipeline run? Seems like magic but should make the index file easy to manage. I will do that.

Reply:
I know... I'm just curious how the S3 output stream manages the upload under the hood.